### PR TITLE
Add prefetch buffer to RandomAccessFileReader

### PR DESCRIPTION
**Pull request by maysamyabandeh**
The prefetch buffer will be used with direct io so that reading sequential metadata does not cause many small IOs. Currently prefetch is disabled with direct io. It is enabled in ReadaheadRandomAccessFile but due to thread safety concerns, this augmented RandomAccessFile is not in table cache to be used for normal file reads.

The prefetch buffer is not thread-safe and meant to be only be used during initialization phase when only one thread is reading some metadata. It is created explicitly upon request and must be deleted explicitly before the pointer is shared with other threads.

The current impl is very simple, limiting the buffer to only prefetched data. and not that of normal reads. This simple impl is sufficient for metadata reads. If this buffer meant to replace ReadaheadRandomAccessFile, we can improve it to also cache the reads.

The current impl has the problem of double buffering if the underlying RandomAccessFile also caches the read data, like in ReadaheadRandomAccessFile. However this problem should not cause much trouble since the prefetch buffer is only used during the short time of initialization. We can still avoid double buffering by adding an API for RandomAccessFile that tells whether its impl does any kind of buffering.

### CODE REVIEW

File: util/file_reader_writer.h [hunk 1]
```diff
@@ -105,7 +125,35 @@ class RandomAccessFileReader {
...
   Status Prefetch(uint64_t offset, size_t n) const {
-    return file_->Prefetch(offset, n);
+    if (mutual_access_ && use_direct_io()) {
+      size_t prefetch_offset = TruncateToPageBoundary(alignment_, offset);
+      if (offset >= buffer_offset_ &&
+          offset + n <= buffer_offset_ + buffer_len_) {
+        return Status::OK();
+      }
+      return ReadIntoPrefetchBuffer(
+          prefetch_offset, Roundup(offset + n, alignment_) - prefetch_offset);
```
Comment (siying) [inline util/file_reader_writer.h]: I still feel uncomfortable that how it complicates the thread-safety guarantee of this class. Now we aren't able to say RandomAccessFileReader::Read() is thread safe. The thread-safety model becomes, we have a switch MarkForConcurrentAccess() to see whether it is thread safe or not.

**Comment by maysamyabandeh** [comment]
Thanks @siying for the review. This approach has the advantage of keeping the file API and hence avoid the complexity of duplicating the code that reads from the file and the code that reads from the buffer. You can see how this simplifies the pending PR (https://github.com/facebook/rocksdb/pull/2661) that requires prefetching metadata.

As for thread-safety concerns you are right about it getting less clear. But note that this is a temporary approach until we have @sagar0's proper fix by the end of H2.

I actually do not quite follow your comment about doing it before compressed block cache. Which is the link in the code that you suggest to change?

**Comment by siying** [comment]
@maysamyabandeh here is the persistent cache code: https://github.com/facebook/rocksdb/blob/master/table/format.cc#L342-L390

This is just before ReadBlock(). The first thing of ReadBlock() is the read the block out: https://github.com/facebook/rocksdb/blob/master/table/format.cc#L274 and this is the only place we use the RandomAccessFileReader. If things are prefetched in the BlockBasedTableReader::Open() itself, the prefetched data can be passed, so that we can directly go to the uncompression part starting from here: https://github.com/facebook/rocksdb/blob/master/table/format.cc#L407-L435 . We don't need to touch RandomAccessFileReader at all.

**Comment by maysamyabandeh** [comment]
Abandoning this in favor of https://github.com/facebook/rocksdb/pull/2708

**Pull request closed without merging**